akamai/cache-client.go

```diff
@@ -190,77 +176,65 @@ func (cpc *CachePurgeClient) purge(urls []string) error {
+// along with an Akamai authorization header.
+func (cpc *CachePurgeClient) authedRequest(endpoint string, body v3PurgeRequest) error {
+	reqBody, err := json.Marshal(body)
 	if err != nil {
-		return errFatal(err.Error())
+		return fmt.Errorf("%s: %w", err, errFatal)
 	}
-	req, err := http.NewRequest(
-		"POST",
-		target,
-		bytes.NewBuffer(reqJSON),
-	)
+
+	req, err := http.NewRequest("POST", endpoint, bytes.NewBuffer(reqBody))
 	if err != nil {
-		return errFatal(err.Error())
+		return fmt.Errorf("%s: %w", err, errFatal)
 	}
 
-	parsedTarget, err := url.Parse(target)
+	endpointURL, err := url.Parse(endpoint)
 	if err != nil {
-		return errFatal(fmt.Sprintf("parsing url %q: %s", target, err))
+		return fmt.Errorf("while parsing %q as URL: %s: %w", endpoint, err, errFatal)
 	}
 
-	// Create authorization header for request
-	authHeader, err := cpc.constructAuthHeader(
-		reqJSON,
-		parsedTarget.Path,
-		core.RandomString(16),
-	)
+	authorization, err := cpc.makeAuthHeader(reqBody, endpointURL.Path, core.RandomString(16))
 	if err != nil {
-		return errFatal(err.Error())
+		return fmt.Errorf("%s: %w", err, errFatal)
 	}
-	req.Header.Set("Authorization", authHeader)
+	req.Header.Set("Authorization", authorization)
 	req.Header.Set("Content-Type", "application/json")
+	cpc.log.Debugf("POSTing to %q with header %q, and body %q", endpoint, authorization, reqBody)
 
-	cpc.log.Debugf("POSTing to %s with Authorization %s: %s",
-		target, authHeader, reqJSON)
-
-	s := cpc.clk.Now()
+	start := cpc.clk.Now()
 	resp, err := cpc.client.Do(req)
-	cpc.purgeLatency.Observe(cpc.clk.Since(s).Seconds())
+	cpc.purgeLatency.Observe(cpc.clk.Since(start).Seconds())
 	if err != nil {
 		return err
 	}
+	defer resp.Body.Close()
+
 	if resp.Body == nil {
-		return fmt.Errorf("No response body")
+		return fmt.Errorf("response body was empty")
 	}
-	body, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		_ = resp.Body.Close()
-		return err
-	}
-	err = resp.Body.Close()
+
+	respBody, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
 		return err
 	}
 
-	// Check purge was successful
+	// Ensure that the purge request was successful.
 	var purgeInfo purgeResponse
-	err = json.Unmarshal(body, &purgeInfo)
+	err = json.Unmarshal(respBody, &purgeInfo)
 	if err != nil {
-		return fmt.Errorf("response to %s: json.Unmarshal(%q): %w", target, body, err)
+		return fmt.Errorf("while unmarshalling response %q from endpoint %q as JSON: %w", respBody, endpoint, err)
 	}
+
 	if purgeInfo.HTTPStatus != http.StatusCreated || resp.StatusCode != http.StatusCreated {
 		if purgeInfo.HTTPStatus == http.StatusForbidden {
-			return errFatal("Unauthorized to purge URLs")
+			return fmt.Errorf("the client was not authorized to make this request: %w", errFatal)
 		}
-		return fmt.Errorf("Unexpected HTTP status code '%d': %s", resp.StatusCode, string(body))
+		return fmt.Errorf("received unexpected HTTP response code '%d', for body %q", resp.StatusCode, respBody)
```

Comment:
As long as we're bikeshedding, I'm going to posit a rule that just occurred to me: We can usually make error messages shorter and clearer by removing "unexpected" (after all, it's an error!):

`HTTP %d (body %q)`.

Also, I'm a big fan of adding the URL to all error messages from the HTTP library, since it doesn't add those itself. Usually I put the URL-adding at a level above the status-code checking, because it should apply to both status code errors and transport errors. I don't see such a URL-added `fmt.Errorf` in here - want to add it? I consider it optional for this cleanup.

Reply:
I believe I've made the suggested edits here. Please let me know if I've misunderstood your comment.